wallet-api/src/Wallet/Emulator/AddressMap.hs

```diff
@@ -34,23 +36,26 @@ instance Monoid AddressMap where
     mappend = (<>)
     mempty = empty
 
--- | The unspent transaction outputs at an address
-getUtxoAt :: Address' -> AddressMap -> Map TxOutRef' Value
-getUtxoAt addr = Map.findWithDefault Map.empty addr . getAddressMap
+type instance Index AddressMap = Address'
+type instance IxValue AddressMap = Map TxOutRef' Value
 
--- | Set the funds at an address to a value
-setUtxoAt :: Address' -> Map TxOutRef' Value -> AddressMap -> AddressMap
-setUtxoAt adr mp = AddressMap . Map.insert adr mp . getAddressMap
+instance Ixed AddressMap where
+    ix adr f (AddressMap mp) = AddressMap <$> ix adr f mp
+
+instance At AddressMap where
+    at idx = lens g s where
+        g (AddressMap mp) = mp ^. at idx
+        s (AddressMap mp) utxo = AddressMap $ mp & at idx .~ utxo
 
 -- | Empty [[AddressMap]]
 empty :: AddressMap
```

Comment:
Er, this is what's looping. I meant just export `mempty` and don't bother with a separate `empty` binding.

Reply:
I am blind 🙈